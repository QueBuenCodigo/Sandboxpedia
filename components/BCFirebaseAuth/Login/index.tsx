import React, { useState } from "react";
import styled from "styled-components";
import { FaFacebookF, FaTwitter } from "react-icons/fa";
import { Formik, Form } from "formik";
import * as yup from "yup";
import FieldFormItem from "../BCField";
import BCButton from "../BCButton";
import authHelper from "../utils/authHelper";
import {Step} from "../"
import {errors} from "../constants/errors"

interface Props{
  signInSuccessWithAuthResult:()=>void;
  signInErrorWithErrorMessage:(errorMessage:string)=>void;
  goTo:(step:Step)=>void;
}

const Login:React.FC<Props> = (props) => {
  const {signInSuccessWithAuthResult,signInErrorWithErrorMessage, goTo} = props;

  const [error, setError] = useState({});
  const [loading, setLoading] = useState(false);

  const initialValues = {
    email: "",
    password: "",
  };

  const validationSchema = yup.object({
    email: yup
      .string()
      .email("Revisa qué tu correo esté bien escrito")
      .required("Falta el correo electrónico."),
    password: yup.string().required("Falta la contraseña."),
  });

  const handleFacebookAuth = async () => {
    const provider = "facebook";
    await authHelper.login(provider);
  };

  const handleTwitterAuth = async () => {
    const provider = "twitter";
    await authHelper.login(provider);
  };

  const handleEmailAuth = async () => {
    const provider = "password";
    const { user, error } = await authHelper.login(
      provider,
      "eduardo@eduardoricardez.com",
      "121232"
    );
    if (error) {
      signInErrorWithErrorMessage(errors[error.code]);
    }
  };

  return (
    <LoginView>
      <LoginCard>
        <LoginContainer>
          <LoginSocial>
            <Facebook onClick={handleFacebookAuth}>
              <FaFacebookF />
              <span>Inicia sesión con Facebook</span>
            </Facebook>
            <Twitter onClick={handleTwitterAuth}>
              <FaTwitter />
              <span>Inicia sesión con Twitter</span>
            </Twitter>
          </LoginSocial>
          <LoginDivider>
            <LoginDividerText>o</LoginDividerText>
          </LoginDivider>
          <LoginWithEmail>
            <Formik
              initialValues={initialValues}
              // validationSchema={validationSchema}
              onSubmit={handleEmailAuth}
            >
              {() => (
                <StyledForm>
                  <LoginInput name="email" placeholder="Tu email" label="Tu email" />
                  <LoginInput name="password" placeholder="Tu contraseña" label="Tu contraseña" />
                  <LoginButton type="submit">Inicia sesión</LoginButton>
                  <LostPassword onClick={()=>goTo(Step.lostPassword)}>¿Olvidaste tu contraseña?</LostPassword>
                </StyledForm>
              )}
            </Formik>
          </LoginWithEmail>
          <AccountFooter>
            <span>¿Aún no tienes cuenta en Buena Tienda?</span>

            <RegisterButton onClick={()=>goTo(Step.signup)}>Regístrate</RegisterButton>
          </AccountFooter>
        </LoginContainer>
      </LoginCard>
    </LoginView>
  );
};

const StyledForm = styled(Form)`
  display: grid;
  grid-template-columns: 100%;
  gap:16px;
`;

const LoginView = styled.div`
  @media only screen and (min-width: 48em) {
    margin: 2rem 0;
  }
`;

const LoginCard = styled.div`
  grid-template-columns: minmax(auto, 405px);
  text-align: center;
  @media only screen and (min-width: 48em) {
    grid-template-columns: minmax(auto, 376px);
    display: grid;
    -webkit-box-pack: center;
    -moz-box-pack: center;
    -o-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
  }
`;

const LoginContainer = styled.div`
  margin-bottom: 1.5rem;
  @media only screen and (min-width: 48em) {
    background-color: #fff;
    width: 376px;
    margin: 0 auto 1.5rem auto;
    -webkit-box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
    box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
  }
`;

const LoginSocial = styled.div`
  padding: 16px 10px 0 10px;
  @media only screen and (min-width: 48em) {
    padding: 24px 40px 0px 40px;
  }
  @media only screen and (min-width: 64em) {
    padding: 24px 32px 0px 32px;
  }

  button {
    display: block;
    margin-bottom: 10px;
  }
`;

const Button = styled.button`
  width: 100%;
  background-color: #d3d3d3;
  border: none;
  -webkit-border-radius: 3px;
  border-radius: 3px;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  color: #fff;
  cursor: pointer;
  display: inline-block;
  font-family: "Lato", sans-serif;
  font-size: 13px;
  padding: 8px 0.8em 6px 0.8em;
  text-decoration: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  user-select: none;
  -webkit-transition: 0.2s;
  -moz-transition: 0.2s;
  -o-transition: 0.2s;
  -ms-transition: 0.2s;
  transition: 0.2s;
  vertical-align: middle;
  white-space: nowrap;
  overflow: hidden;
  -o-text-overflow: ellipsis;
  text-overflow: ellipsis;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: 0;
  &:active {
    -webkit-transform: scale(0.9);
    -moz-transform: scale(0.9);
    -o-transform: scale(0.9);
    -ms-transform: scale(0.9);
    transform: scale(0.9);
  }
  svg {
    margin-right: 0.5em;
    vertical-align: middle;
  }
`;

const Facebook = styled(Button)`
  font-size: 18px;
  padding: 11px 0.8em 12px 0.8em;
  background: -webkit-linear-gradient(
    right,
    #3b5999 0%,
    #3b5999 50%,
    #4e69a3 100%
  );
  background: -moz-linear-gradient(
    right,
    #3b5999 0%,
    #3b5999 50%,
    #4e69a3 100%
  );
  background: -o-linear-gradient(right, #3b5999 0%, #3b5999 50%, #4e69a3 100%);
  background: -ms-linear-gradient(right, #3b5999 0%, #3b5999 50%, #4e69a3 100%);
  background: linear-gradient(to left, #3b5999 0%, #3b5999 50%, #4e69a3 100%);
  -webkit-box-shadow: 0 1px 1px 0 #355089;
  box-shadow: 0 1px 1px 0 #355089;
  &:hover {
    -webkit-box-shadow: 0 2px 13px 0 rgba(53, 80, 137, 0.5);
    box-shadow: 0 2px 13px 0 rgba(53, 80, 137, 0.5);
  }
`;

const Twitter = styled(Button)`
  font-size: 18px;
  padding: 11px 0.8em 12px 0.8em;
  background: -webkit-linear-gradient(
    right,
    #60a9de 0%,
    #60a9de 50%,
    #6fb1e1 100%
  );
  background: -moz-linear-gradient(
    right,
    #60a9de 0%,
    #60a9de 50%,
    #6fb1e1 100%
  );
  background: -o-linear-gradient(right, #60a9de 0%, #60a9de 50%, #6fb1e1 100%);
  background: -ms-linear-gradient(right, #60a9de 0%, #60a9de 50%, #6fb1e1 100%);
  background: linear-gradient(to left, #60a9de 0%, #60a9de 50%, #6fb1e1 100%);
  -webkit-box-shadow: 0 1px 1px 0 #5698c7;
  box-shadow: 0 1px 1px 0 #5698c7;
  &:hover {
    -webkit-box-shadow: 0 2px 13px 0 rgba(86, 152, 199, 0.5);
    box-shadow: 0 2px 13px 0 rgba(86, 152, 199, 0.5);
  }
`;

const LoginDivider = styled.div`
  text-align: center;
  overflow: hidden;
  margin: 1rem auto;
  width: 94%;
  @media only screen and (min-width: 48em) {
    width: 79%;
  }
  @media only screen and (min-width: 64em) {
    width: 82%;
  }
`;

const LoginDividerText = styled.span`
  position: relative;
  padding: 12px;
  font-weight: 600;
  color: #273b47;
  font-size: 20px;
  &::before {
    content: "";
    position: absolute;
    top: 50%;
    right: 100%;
    width: 5000px;
    border-bottom: 1px solid #e4e4e4;
  }
  &::after {
    content: "";
    position: absolute;
    top: 50%;
    left: 100%;
    width: 5000px;
    border-bottom: 1px solid #e4e4e4;
  }
`;

const LoginWithEmail = styled.div`
  padding: 0px 10px 24px;
  @media only screen and (min-width: 48em) {
    padding: 0px 40px 24px;
  }
  @media only screen and (min-width: 64em) {
    padding: 0px 32px 24px;
  }
`;

const LoginInput = styled(FieldFormItem)`
  margin: 0;
  text-align: left;
  input {
    padding: 15px 12px;
    border-radius: 0.25rem;
  }
`;

const LoginButton = styled(Button)`
  display: block;
  width: 100%;
  padding: 14px 0;
  font-weight: bold;
  font-size: 1rem;
  background: -webkit-linear-gradient(
    right,
    #95ca3e 0%,
    #95ca3e 50%,
    #85c638 100%
  );
  background: -moz-linear-gradient(
    right,
    #95ca3e 0%,
    #95ca3e 50%,
    #85c638 100%
  );
  background: -o-linear-gradient(right, #95ca3e 0%, #95ca3e 50%, #85c638 100%);
  background: -ms-linear-gradient(right, #95ca3e 0%, #95ca3e 50%, #85c638 100%);
  background: linear-gradient(to left, #95ca3e 0%, #95ca3e 50%, #85c638 100%);
  -webkit-box-shadow: 0 1px 1px 0 #58902d;
  box-shadow: 0 1px 1px 0 #58902d;
  &:hover {
    -webkit-box-shadow: 0 2px 13px 0 rgba(88, 144, 45, 0.49);
    box-shadow: 0 2px 13px 0 rgba(88, 144, 45, 0.49);
    color: #fff;
  }
`;

const LostPassword = styled(BCButton)`
  margin-top: 16px;
  font-size: 14px;
  color: #0791e6;
  text-decoration: none;
  display: block;
`;

const AccountFooter = styled.div`
  padding: 24px 10px;
  background-color: #f6f6f6;
  @media only screen and (min-width: 48em) {
    padding: 24px 42px;
  }
  @media only screen and (min-width: 64em) {
    padding: 24px 32px;
  }
  span {
    display: block;
    margin-bottom: 16px;
    font-size: 14px;
    font-weight: bold;
    line-height: 1.37;
    color: #3a3c3f;
  }
`;

const RegisterButton = styled.button`
  display: block;
  font-size: 16px;
  padding: 12px 0.8em 12px 0.8em;
  font-size: 1rem;
  border: 1px solid #0791e6;
  -webkit-border-radius: 4px;
  border-radius: 4px;
  color: #0791e6;
  cursor: pointer;
  outline: none;
  background: #fff;
  text-decoration: none;
  font-weight: bold;
  width: 100%;
`;

export default Login;
