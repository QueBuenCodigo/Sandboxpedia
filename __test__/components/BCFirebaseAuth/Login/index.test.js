import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';
import Login from '../../../../components/BCFirebaseAuth/Login/index';

describe('<Login/>',()=>{

    const login=mount(<Login/>)

    test('test del renderizado de login',()=>{

        expect(login.length).toEqual(1);
    });
});


describe('Login Snapshot',()=>{

    test('comprobar la UI del componente Login',()=>{
        const login= create(<Login/>);
        expect(login.toJSON()).toMatchSnapshot();
    });

});
