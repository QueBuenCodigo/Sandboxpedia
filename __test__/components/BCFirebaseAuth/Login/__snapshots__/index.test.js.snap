// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Snapshot comprobar la UI del componente Login 1`] = `
<div
  className="Login__LoginView-sc-1wqjwyn-1 iWKuiH"
>
  <div
    className="Login__LoginCard-sc-1wqjwyn-2 bbKUTC"
  >
    <div
      className="Login__LoginContainer-sc-1wqjwyn-3 kpOQgL"
    >
      <div
        className="Login__LoginSocial-sc-1wqjwyn-4 eMdmsL"
      >
        <button
          className="Login__Button-sc-1wqjwyn-5 Login__Facebook-sc-1wqjwyn-6 bnQKRu"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
            />
          </svg>
          <span>
            Inicia sesión con Facebook
          </span>
        </button>
        <button
          className="Login__Button-sc-1wqjwyn-5 Login__Twitter-sc-1wqjwyn-7 epOzkN"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
            />
          </svg>
          <span>
            Inicia sesión con Twitter
          </span>
        </button>
      </div>
      <div
        className="Login__LoginDivider-sc-1wqjwyn-8 jeMCSc"
      >
        <span
          className="Login__LoginDividerText-sc-1wqjwyn-9 dafOoy"
        >
          o
        </span>
      </div>
      <div
        className="Login__LoginWithEmail-sc-1wqjwyn-10 nWTls"
      >
        <form
          action="#"
          className="Login__StyledForm-sc-1wqjwyn-0 hNMcuq"
          onReset={[Function]}
          onSubmit={[Function]}
        >
          <div
            className="BCField__FormItem-sc-85pqlt-0 dvOPuG"
          >
            <label
              className="BCField__InputLabel-sc-85pqlt-1 iLkEPd form__label"
            >
              <input
                className="BCField__Input-sc-85pqlt-2 bbKpRK form__field"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Tu email"
                value=""
              />
              <span>
                Tu email
              </span>
            </label>
          </div>
          <div
            className="BCField__FormItem-sc-85pqlt-0 dvOPuG"
          >
            <label
              className="BCField__InputLabel-sc-85pqlt-1 iLkEPd form__label"
            >
              <input
                className="BCField__Input-sc-85pqlt-2 bbKpRK form__field"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Tu contraseña"
                value=""
              />
              <span>
                Tu contraseña
              </span>
            </label>
          </div>
          <button
            className="Login__Button-sc-1wqjwyn-5 Login__LoginButton-sc-1wqjwyn-12 hamVkz"
            type="submit"
          >
            Inicia sesión
          </button>
          <button
            className="BCButton__Button-sc-18tvwvx-0 lgdhbd Login__LostPassword-sc-1wqjwyn-13 gJFSQE"
            onClick={[Function]}
          >
            ¿Olvidaste tu contraseña?
          </button>
        </form>
      </div>
      <div
        className="Login__AccountFooter-sc-1wqjwyn-14 gqEwqS"
      >
        <span>
          ¿Aún no tienes cuenta en Buena Tienda?
        </span>
        <button
          className="Login__RegisterButton-sc-1wqjwyn-15 jBgDsH"
          onClick={[Function]}
        >
          Regístrate
        </button>
      </div>
    </div>
  </div>
</div>
`;
